File: ios/pixel_compressor/Sources/pixel_compressor/ImageEngine.swift
import CoreGraphics
import Foundation
import ImageIO

/// Implements `ImageHostApi.compressImage`. `CGImage`/`CGImageSource`/
/// `CGImageDestination` only — no UIKit — so this file works unmodified on
/// macOS via the repo's manual-sync copy.
final class ImageEngine {
  private let cacheManager: CacheManager
  private let taskRegistry: TaskRegistry
  private let progressReporter: ProgressReporter

  private enum Constants {
    static let maxAttempts = 8
    static let qualityStep = 15
    static let qualityFloor = 10
    static let resolutionStepFactor = 0.75
    static let minEdgePx = 200
  }

  init(cacheManager: CacheManager, taskRegistry: TaskRegistry, progressReporter: ProgressReporter) {
    self.cacheManager = cacheManager
    self.taskRegistry = taskRegistry
    self.progressReporter = progressReporter
  }

  func compress(request: ImageCompressRequest) throws -> CompressionResultMessage {
    let taskId = request.taskId
    let startTime = Date()

    try checkCancelled(taskId)
    progressReporter.reportProgress(taskId: taskId, stage: .preparing, percent: 0)

    guard FileManager.default.fileExists(atPath: request.sourcePath) else {
      throw PixelCompressorError.invalidMedia("Source file does not exist: \(request.sourcePath)")
    }
    let sourceURL = URL(fileURLWithPath: request.sourcePath)

    guard let source = CGImageSourceCreateWithURL(sourceURL as CFURL, nil) else {
      throw PixelCompressorError.decodingError("Could not open source image")
    }
    guard CGImageSourceGetType(source) != nil, CGImageSourceGetCount(source) > 0 else {
      throw PixelCompressorError.unsupportedMedia("Source file isn't a readable image")
    }

    let sourceProperties =
      (CGImageSourceCopyPropertiesAtIndex(source, 0, nil) as? [CFString: Any]) ?? [:]
    let sourceOrientation = (sourceProperties[kCGImagePropertyOrientation] as? Int) ?? 1
    let sourceOrientationDegrees =
      request.autoCorrectOrientation
      ? RotationMath.degrees(forExifOrientation: sourceOrientation)
      : 0
    let requestedRotation = try RotationMath.requireCardinal(Int(request.rotationDegrees))

    let originalSize =
      (try? FileManager.default.attributesOfItem(atPath: request.sourcePath)[.size] as? Int64) ?? 0

    let sourceExtension = (sourceURL.pathExtension.isEmpty ? "jpg" : sourceURL.pathExtension)
    let outputFormat = request.format ?? ImageFormatCodec.format(forPathExtension: sourceExtension)
    // WebP output is rejected up front, before any decode work, per the
    // hard platform constraint (ImageIO can decode WebP, never encode it).
    _ = try ImageFormatCodec.utType(for: outputFormat)

    try checkCancelled(taskId)
    progressReporter.reportProgress(taskId: taskId, stage: .decoding, percent: 10)

    let requestedMaxWidth = request.maxWidth.map { Int($0) }
    let requestedMaxHeight = request.maxHeight.map { Int($0) }
    let requestedMaxEdge: Int? = {
      switch (requestedMaxWidth, requestedMaxHeight) {
      case let (.some(w), .some(h)): return max(w, h)
      case let (.some(w), .none): return w
      case let (.none, .some(h)): return h
      case (.none, .none): return nil
      }
    }()

    let baseImage = try decodeAndOrient(
      source: source,
      sourceOrientationDegrees: sourceOrientationDegrees,
      requestedRotation: requestedRotation,
      maxEdge: requestedMaxEdge
    )

    let outputURL = try resolveOutputURL(request: request, format: outputFormat)
    try? FileManager.default.removeItem(at: outputURL)

    progressReporter.reportProgress(taskId: taskId, stage: .encoding, percent: 40)

    let properties = exifOutputProperties(
      policy: request.exifPolicy, sourceProperties: sourceProperties)

    if let targetSizeBytes = request.targetSizeBytes {
      try compressToTargetSize(
        taskId: taskId,
        baseImage: baseImage,
        format: outputFormat,
        initialQuality: Int(request.quality),
        targetSizeBytes: Int(targetSizeBytes),
        properties: properties,
        outputURL: outputURL
      )
    } else {
      try ImageFormatCodec.encode(
        image: baseImage,
        to: outputURL,
        format: outputFormat,
        quality: Int(request.quality),
        properties: properties
      )
    }

    try checkCancelled(taskId)
    progressReporter.reportProgress(taskId: taskId, stage: .finalizing, percent: 95)

    guard
      let outputAttrs = try? FileManager.default.attributesOfItem(atPath: outputURL.path),
      let outputSize = outputAttrs[.size] as? Int64,
      outputSize > 0
    else {
      throw PixelCompressorError.encodingError("Output file is empty or missing")
    }

    let durationMs = Int64(Date().timeIntervalSince(startTime) * 1000)
    let ratio = originalSize > 0 ? Double(outputSize) / Double(originalSize) : 0

    progressReporter.reportProgress(taskId: taskId, stage: .completed, percent: 100, force: true)

    return CompressionResultMessage(
      outputPath: outputURL.path,
      originalSizeBytes: originalSize,
      outputSizeBytes: outputSize,
      compressionRatio: ratio,
      durationMs: durationMs,
      codec: ImageFormatCodec.formatName(for: outputFormat),
      format: ImageFormatCodec.formatName(for: outputFormat)
    )
  }

  /// Decodes the source into a single upright `CGImage` with rotation
  /// baked in and (optionally) downscaled to fit `maxEdge` on its longest
  /// side.
  ///
  /// When downscaling is requested, uses `CGImageSourceCreateThumbnailAtIndex`
  /// with `kCGImageSourceCreateThumbnailWithTransform`, which bakes the
  /// source's own EXIF orientation automatically during the resize — so
  /// only the caller's own `requestedRotation` still needs applying on top,
  /// with a manual `CGContext` pass (applying `sourceOrientationDegrees`
  /// again here would double-rotate).
  ///
  /// When no downscaling is requested, the full-resolution image is
  /// decoded raw (no orientation baked by ImageIO) and the combined
  /// source+requested rotation is applied in one manual pass.
  private func decodeAndOrient(
    source: CGImageSource,
    sourceOrientationDegrees: Int,
    requestedRotation: Int,
    maxEdge: Int?
  ) throws -> CGImage {
    if let maxEdge = maxEdge {
      let options: [CFString: Any] = [
        kCGImageSourceThumbnailMaxPixelSize: maxEdge,
        kCGImageSourceCreateThumbnailFromImageAlways: true,
        kCGImageSourceCreateThumbnailWithTransform: true,
        kCGImageSourceShouldCacheImmediately: true,
      ]
      guard
        let thumbnail = CGImageSourceCreateThumbnailAtIndex(source, 0, options as CFDictionary)
      else {
        throw PixelCompressorError.decodingError("Could not decode/downscale source image")
      }
      if requestedRotation != 0 {
        return try ImageFormatCodec.rotate(thumbnail, byDegrees: requestedRotation)
      }
      return thumbnail
    }

    guard let full = CGImageSourceCreateImageAtIndex(source, 0, nil) else {
      throw PixelCompressorError.decodingError("Could not decode source image")
    }
    let totalRotation = RotationMath.combine(
      sourceDegrees: sourceOrientationDegrees, requestedDegrees: requestedRotation)
    if totalRotation != 0 {
      return try ImageFormatCodec.rotate(full, byDegrees: totalRotation)
    }
    return full
  }

  private func compressToTargetSize(
    taskId: String,
    baseImage: CGImage,
    format: ImageFormatWire,
    initialQuality: Int,
    targetSizeBytes: Int,
    properties: [CFString: Any]?,
    outputURL: URL
  ) throws {
    var quality = max(Constants.qualityFloor, min(initialQuality, 100))
    var image = baseImage

    var attempt = 0
    while attempt < Constants.maxAttempts {
      try checkCancelled(taskId)
      progressReporter.reportProgress(
        taskId: taskId, stage: .encoding, percent: 40 + Double(attempt) * 6,
        note: "attempt \(attempt + 1), quality \(quality), edge \(max(image.width, image.height))"
      )

      try ImageFormatCodec.encode(
        image: image, to: outputURL, format: format, quality: quality, properties: properties)

      let size =
        (try? FileManager.default.attributesOfItem(atPath: outputURL.path)[.size] as? Int64) ?? 0
      if size > 0 && size <= targetSizeBytes {
        return
      }

      attempt += 1
      if quality > Constants.qualityFloor {
        quality = max(Constants.qualityFloor, quality - Constants.qualityStep)
      } else {
        let currentEdge = max(image.width, image.height)
        let newEdge = max(
          Int(Double(currentEdge) * Constants.resolutionStepFactor), Constants.minEdgePx)
        if newEdge >= currentEdge {
          break
        }
        image = try ImageFormatCodec.resize(image, maxPixelSize: newEdge)
      }
    }

    throw PixelCompressorError.targetSizeUnachievable(
      "Could not reach target size of \(targetSizeBytes) bytes for image after \(Constants.maxAttempts) attempts"
    )
  }

  /// `ExifPolicy.keep`: copies the source's EXIF/TIFF/GPS property
  /// dictionaries as-is but forces orientation to 1 — pixels have already
  /// been rotated upright by `decodeAndOrient`, so a stale orientation tag
  /// would double-rotate the image in viewers that respect it.
  /// `ExifPolicy.strip`: writes no metadata at all.
  private func exifOutputProperties(
    policy: ExifPolicyWire, sourceProperties: [CFString: Any]
  ) -> [CFString: Any]? {
    switch policy {
    case .strip:
      return nil
    case .keep:
      var properties = sourceProperties
      properties[kCGImagePropertyOrientation] = 1
      return properties
    }
  }

  private func resolveOutputURL(request: ImageCompressRequest, format: ImageFormatWire) throws
    -> URL
  {
    if let outputPath = request.outputPath, !outputPath.isEmpty {
      return URL(fileURLWithPath: outputPath)
    }
    let fileName = "\(request.taskId).\(ImageFormatCodec.fileExtension(for: format))"
    let path = try cacheManager.resolveOutputPath(kind: .images, fileName: fileName)
    return URL(fileURLWithPath: path)
  }

  private func checkCancelled(_ taskId: String) throws {
    if taskRegistry.isCancelled(taskId: taskId) {
      throw PixelCompressorError.cancelled()
    }
  }
}

File: android/src/main/kotlin/com/therivanta/pixelcompressor/image/ImageEngine.kt
package com.therivanta.pixelcompressor.image

import android.content.Context
import com.therivanta.pixelcompressor.CompressionResultMessage
import com.therivanta.pixelcompressor.CompressionStageWire
import com.therivanta.pixelcompressor.ExifPolicyWire
import com.therivanta.pixelcompressor.ImageCompressRequest
import com.therivanta.pixelcompressor.ImageFormatWire
import com.therivanta.pixelcompressor.cache.CacheKind
import com.therivanta.pixelcompressor.cache.CacheManager
import com.therivanta.pixelcompressor.capability.CapabilityProbe
import com.therivanta.pixelcompressor.errors.PixelCompressorErrors
import com.therivanta.pixelcompressor.progress.ProgressReporter
import com.therivanta.pixelcompressor.util.RotationMath
import java.io.File
import java.io.IOException
import kotlin.math.max
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive

/** Image compression pipeline: decode (EXIF-orientation-aware, downsampled)
 * -> rotate -> resize -> encode (JPEG/PNG/WEBP via `Bitmap.compress`, HEIC
 * via [HeicWriterSupport]) -> EXIF keep/strip -> optional target-size
 * iterative search.
 */
class ImageEngine(
  private val context: Context,
  private val cacheManager: CacheManager,
  private val capabilityProbe: CapabilityProbe,
  private val progressReporter: ProgressReporter,
) {

  private companion object {
    const val MAX_ATTEMPTS = 8
    const val QUALITY_STEP = 15
    const val QUALITY_FLOOR = 10
    const val RESOLUTION_STEP_FACTOR = 0.75
    const val MIN_EDGE_PX = 200
    const val INITIAL_TARGET_QUALITY = 90
  }

  suspend fun compress(request: ImageCompressRequest): CompressionResultMessage {
    val startTime = System.currentTimeMillis()
    val taskId = request.taskId
    progressReporter.onProgress(taskId, CompressionStageWire.PREPARING, 0.0)

    val sourceFile = File(request.sourcePath)
    if (!sourceFile.exists()) {
      throw PixelCompressorErrors.invalidMedia("Source file does not exist: ${request.sourcePath}")
    }
    if (!sourceFile.canRead()) {
      throw PixelCompressorErrors.permissionDenied("Cannot read source file: ${request.sourcePath}")
    }
    val originalSizeBytes = sourceFile.length()

    // Format + codec-capability check happens before any decode work, per
    // the "never silently substitute a codec" contract.
    val format = request.format ?: ImageFormatCodec.detectFromPath(request.sourcePath)
    if (format == ImageFormatWire.HEIC && !capabilityProbe.isHevcSupported()) {
      throw PixelCompressorErrors.unsupportedCodec(
        "HEIC encode requires a hardware HEVC encoder, which this device does not report support for",
      )
    }

    currentCoroutineContext().ensureActive()
    progressReporter.onProgress(taskId, CompressionStageWire.ANALYZING, 5.0)

    try {
      BitmapTransform.decodeBounds(request.sourcePath)
    } catch (e: Exception) {
      throw PixelCompressorErrors.unsupportedMedia(
        "Source is not a decodable image: ${request.sourcePath}",
        e.toString(),
      )
    }

    val sourceOrientationDegrees = if (request.autoCorrectOrientation) {
      ExifSupport.readOrientationDegrees(request.sourcePath)
    } else {
      0
    }
    val totalRotation = RotationMath.combineRotation(sourceOrientationDegrees, request.rotationDegrees.toInt())
    val requestedMaxWidth = request.maxWidth?.toInt()
    val requestedMaxHeight = request.maxHeight?.toInt()

    progressReporter.onProgress(taskId, CompressionStageWire.DECODING, 10.0)

    val scratchDir = File(context.cacheDir, "pixel_compressor/tmp").apply { mkdirs() }

    val quality: Int
    val encodedBytes: ByteArray

    if (request.targetSizeBytes == null) {
      val bitmap = decodeRotateFit(request.sourcePath, totalRotation, requestedMaxWidth, requestedMaxHeight)
      currentCoroutineContext().ensureActive()
      progressReporter.onProgress(taskId, CompressionStageWire.ENCODING, 50.0)
      quality = request.quality.toInt().coerceIn(1, 100)
      encodedBytes = try {
        ImageFormatCodec.encode(bitmap, format, quality, scratchDir)
      } finally {
        bitmap.recycle()
      }
    } else {
      val result = compressToTargetSize(
        taskId = taskId,
        sourcePath = request.sourcePath,
        totalRotation = totalRotation,
        requestedMaxWidth = requestedMaxWidth,
        requestedMaxHeight = requestedMaxHeight,
        format = format,
        targetSizeBytes = request.targetSizeBytes,
        scratchDir = scratchDir,
      )
      quality = result.quality
      encodedBytes = result.bytes
    }

    currentCoroutineContext().ensureActive()
    progressReporter.onProgress(taskId, CompressionStageWire.FINALIZING, 95.0)

    val outputPath = request.outputPath
      ?: cacheManager.resolveOutputPath(CacheKind.IMAGES, ImageFormatCodec.extensionFor(format))
    val outputFile = File(outputPath)
    outputFile.parentFile?.mkdirs()
    try {
      outputFile.writeBytes(encodedBytes)
    } catch (e: IOException) {
      throw mapWriteError(e)
    }

    if (request.exifPolicy == ExifPolicyWire.KEEP) {
      ExifSupport.copyExif(request.sourcePath, outputPath)
    }

    val outputSizeBytes = outputFile.length()
    val durationMs = System.currentTimeMillis() - startTime
    val ratio = if (originalSizeBytes > 0) outputSizeBytes.toDouble() / originalSizeBytes else 0.0

    progressReporter.onProgress(taskId, CompressionStageWire.COMPLETED, 100.0)

    return CompressionResultMessage(
      outputPath = outputPath,
      originalSizeBytes = originalSizeBytes,
      outputSizeBytes = outputSizeBytes,
      compressionRatio = ratio,
      durationMs = durationMs,
      codec = ImageFormatCodec.wireNameFor(format),
      format = ImageFormatCodec.wireNameFor(format),
    )
  }

  private fun decodeRotateFit(
    path: String,
    totalRotation: Int,
    maxWidth: Int?,
    maxHeight: Int?,
  ): android.graphics.Bitmap {
    // Generous pre-rotation sample target: rotation may swap width/height,
    // so sample against the larger of the two requested bounds with margin,
    // then fit exactly post-rotation.
    val sampleEdge = if (maxWidth != null || maxHeight != null) {
      max(maxWidth ?: 0, maxHeight ?: 0) * 2
    } else {
      null
    }
    var bitmap = BitmapTransform.decodeSampledBitmap(path, sampleEdge)
    bitmap = BitmapTransform.rotate(bitmap, totalRotation)
    bitmap = BitmapTransform.fitToBox(bitmap, maxWidth, maxHeight)
    return bitmap
  }

  private data class TargetSizeAttemptResult(
    val bytes: ByteArray,
    val quality: Int,
  )

  private suspend fun compressToTargetSize(
    taskId: String,
    sourcePath: String,
    totalRotation: Int,
    requestedMaxWidth: Int?,
    requestedMaxHeight: Int?,
    format: ImageFormatWire,
    targetSizeBytes: Long,
    scratchDir: File,
  ): TargetSizeAttemptResult {
    var quality = INITIAL_TARGET_QUALITY
    var currentMaxWidth = requestedMaxWidth
    var currentMaxHeight = requestedMaxHeight
    var lastBytes: ByteArray? = null
    var attempt = 0

    while (attempt < MAX_ATTEMPTS) {
      attempt++
      currentCoroutineContext().ensureActive()

      val bitmap = decodeRotateFit(sourcePath, totalRotation, currentMaxWidth, currentMaxHeight)
      val width = bitmap.width
      val height = bitmap.height
      val attemptBytes = try {
        ImageFormatCodec.encode(bitmap, format, quality, scratchDir)
      } finally {
        bitmap.recycle()
      }
      lastBytes = attemptBytes

      val percent = 10.0 + (attempt.toDouble() / MAX_ATTEMPTS) * 80.0
      progressReporter.onProgress(
        taskId,
        CompressionStageWire.ENCODING,
        percent,
        note = "attempt $attempt/$MAX_ATTEMPTS quality=$quality edge=${max(width, height)}",
      )

      if (attemptBytes.size <= targetSizeBytes) {
        return TargetSizeAttemptResult(attemptBytes, quality)
      }

      if (attempt >= MAX_ATTEMPTS) break

      if (quality > QUALITY_FLOOR) {
        quality = max(QUALITY_FLOOR, quality - QUALITY_STEP)
      } else {
        val currentEdge = max(width, height)
        val newEdge = max(MIN_EDGE_PX, (currentEdge * RESOLUTION_STEP_FACTOR).toInt())
        currentMaxWidth = newEdge
        currentMaxHeight = newEdge
      }
    }

    throw PixelCompressorErrors.targetSizeUnachievable(
      "Could not reach target size of $targetSizeBytes bytes after $MAX_ATTEMPTS attempts " +
        "(best attempt: ${lastBytes?.size ?: 0} bytes)",
    )
  }

  private fun mapWriteError(e: IOException): Throwable {
    val message = e.message ?: ""
    return if (message.contains("ENOSPC", ignoreCase = true) ||
      message.contains("No space left", ignoreCase = true)
    ) {
      PixelCompressorErrors.insufficientStorage("No space left to write output file", e.toString())
    } else {
      PixelCompressorErrors.encodingError("Failed to write output file: ${e.message}", e.toString())
    }
  }
}
